File: unityProduct/Assets/Script/ViewController/ViewController.cs
using System.Collections;
using System.Collections.Generic;
using System;
using UnityEngine;
using UnityEngine.UI;
using System.Threading.Tasks;

public class ViewController : MonoBehaviour
{
    static GameObject canvas = null;
    static Dictionary<Type, ViewController> tempUI = new Dictionary<Type, ViewController>();

    public static ViewController GetViewController(Type type)
    {
        if (canvas == null)
        {
            canvas = GameObject.Find("Canvas");
        }

        if (!tempUI.ContainsKey(type))
        {
            if (Config.IS_UI_ASSEST_BUNDLE)
            {
                loadAssestBundle(type);
            }
            else
            {
                loadResource(type);
            }
        }
        return tempUI[type];
    }

    public static T GetViewController<T>() where T : ViewController
    {
        if (canvas == null)
        {
            canvas = GameObject.Find("Canvas");
        }

        Type type = typeof(T);

        if (!tempUI.ContainsKey(type))
        {
            if (Config.IS_UI_ASSEST_BUNDLE)
            {
                loadAssestBundle(type);
            }
            else
            {
                loadResource(type);
            }
        }
        return tempUI[type] as T;
    }


    private static void loadResource(Type type)
    {
        if (tempUI.ContainsKey(type))
            return;

        GameObject viewGO = GameObject.Instantiate(Resources.Load("ui/" + type.ToString())) as GameObject;
        viewGO.transform.SetParent(canvas.transform);
        var rectTransform = viewGO.transform.GetComponent<RectTransform>();
        rectTransform.sizeDelta = new Vector2(1920, 1080);
        rectTransform.anchoredPosition = Vector2.zero;
        tempUI[type] = viewGO.GetComponent<ViewController>();
    }

    private static void loadAssestBundle(Type type)
    {
        if (tempUI.ContainsKey(type))
            return;

        var obj = LoadAssestBundle.Instance.loadAssetBundleSync("ui", type.ToString());
        GameObject viewGO = GameObject.Instantiate(obj) as GameObject;
        viewGO.transform.SetParent(canvas.transform);
        var rectTransform = viewGO.transform.GetComponent<RectTransform>();
        rectTransform.sizeDelta = new Vector2(1920, 1080);
        rectTransform.anchoredPosition = Vector2.zero;
        tempUI[type] = viewGO.GetComponent<ViewController>();
    }

    bool isTryOpen = false;
    private object openData = null;
    private void Awake()
    {
        if (!tempUI.ContainsKey(GetType()))
        {
            tempUI.Add(GetType(), this);
        }
    }

    // Start is called before the first frame update
    void Start()
    {
        init();   
    }

    // Update is called once per frame
    void Update()
    {
        if (isTryOpen)
        {
            isTryOpen = false;
            var input = openData;
            openData = null;
            open(input);
        }
        onTimer();
    }

    protected virtual void init()
    {

    }

    protected virtual void open(object obj)
    {

    }

    public void show()
    {
        this.gameObject.SetActive(true);
        isTryOpen = true;
    }

    public void show(object obj)
    {
        openData = obj;
        show();
    }

    public virtual void close()
    {
        this.gameObject.SetActive(false);
    }

    protected virtual void onTimer()
    {
    }
}

File: FgoProductCocos/assets/script/ViewController/ViewController.ts
import { _decorator, Component, error, find, Node, Prefab, resources,instantiate } from 'cc';
import { Config } from '../Config';
import { ViewControllerManager } from './ViewControllerManager';
const { ccclass, property } = _decorator;

@ccclass('ViewController')
export class ViewController extends Component {

    private static canvas : Node;
    private static viewControllerTemps : Map<Function,ViewController> = new Map();
    
    public static async GetViewController<T extends ViewController>(vcClass: { new(): T }): T
    {
        console.log("vcClass : " + vcClass.name);

        if(!this.viewControllerTemps.has(vcClass))
        {
            if(this.canvas == null)
            {
                this.canvas = find("Canvas");
            }
            if(Config.IS_ASSEST_BUNDLE)
            {

            }
            else
            {
                const path = "ui/" + (vcClass as any).viewResource();
                const node = await this.loadAndInstantiatePrefab(path);
                //const node = await this.loadAndInstantiatePrefab("ui/" + vcClass.name);
            }
        }

        return this.viewControllerTemps.get(vcClass) as T;
    }

    private static async loadAndInstantiatePrefab(path: string): Promise<Node> {
        return new Promise((resolve, reject) => {
            resources.load(path, Prefab, (err, prefab) => {
                if (err || !prefab) {
                    reject(err);
                } else {
                    const node = instantiate(prefab);
                    this.canvas.addChild(node);
                    resolve(node);
                }
            });
        });
    }

    protected static viewResource():string
    {
        return "ViewController";
    }

    private isTryOpen : boolean = false;
    private openData : any = null;
    protected onLoad(): void {
        //console.log(this.constructor.name);
        var ct = this.constructor as Function;
        //console.log("ct : " + ct);
        if(!ViewController.viewControllerTemps.has(ct))
        {
            ViewController.viewControllerTemps.set(ct,this);
        }
    }

    start() {
        this.init();
    }

    update(deltaTime: number) {
        if(this.isTryOpen)
        {
            this.isTryOpen = false;
            var input = this.openData;
            this.openData = null;
            this.open(input);
        }

        this.onTimer();
    }   

    protected init():void
    {

    }

    protected open(obj : any):void
    {

    }

    public close() : void
    {
        this.node.active = false;
    }

    protected onTimer()
    {

    }

    
    public show():void;
    public show(data : any):void;
    public show(data? : any):void
    {
        this.node.active = true;
        this.isTryOpen = true;
        if(data)
        {
            this.openData = data;
        }
    }
}


